snuba/datasets/dataset.py

```diff
@@ -48,6 +49,10 @@ def __init__(
         self.__dataset_schemas = dataset_schemas
         self.__table_writer = table_writer
 
+    @abstractmethod
+    def __str__(self):
+        raise NotImplementedError
```

Comment:
One of my personal gripes with the datasets architecture is that dataset classes are effectively singletons, I don't like that change pushes the design further in that direction. It also seems like it'd be very easy to get this value out of sync with the values defined in `snuba.datasets.factory`, which it seems like it is supposed to correspond to?

It's not a reasonable request to change the dataset architecture right now for the sake of this change, but I think that we should try to make this consistent with the factory code by one of the following:

1. the instantiation code should use this value when instantiation dataset instances instead of the hardcoded names in this file, or
2. this method should use `get_dataset_name` internally, or 
3. this method should be removed and usage replaced with `get_dataset_name`, or
4. this method should be removed and usage replaced with the class name for this method

Reply:
I forgot about get_dataset_name name....

About the singleton, I don't think we have strong assumptions that only one instance per dataset exists. In theory (but I never tried, this may miserably fail) we should be able to remove the dataset cache and instantiate a new one every time it is needed and everything should still work. We do ask people not to assume anything about the instance they get though in query processors.
I guess the issue here may be that returning the name of the dataset is not the responsibility of the __str__ method (which is an instance method), but we should just either use get_dataset_name or have a classmethod that return the name. How does this sound ?

Reply:
I didn't know we have get_dataset_name, let's use that

Reply:
> About the singleton, I don't think we have strong assumptions that only one instance per dataset exists.

Yeah, I think that singleton was the wrong word for me to use here (and this is not the first time I've made this mistake: see https://github.com/getsentry/snuba/pull/696#discussion_r367576977.) Sorry for making that confusing.

I think that my actual issue is more with the fact that a dataset has to be registered _at all_ for some parts of the codebase to function correctly (which is a totally different problem.) In other words, I don't think it is a good thing that some dataset-derived functionality requires that the dataset instance be registered in global scope in order to function correctly.

Using `get_dataset_name` doesn't fix that (and it shouldn't, that is way out of the scope of this change, and I think I've already gone too far off topic here) but is better than introducing that dependency in another way (e.g. as a class or instance method) and it at least makes things consistent with where this dependency already exists.

I do have some fragments of thoughts about how to make this better in the future, I will try to take some time to organize them constructively soon (rather than just complaining about it.)